pandas/core/arrays/datetimes.py

```diff
@@ -1499,6 +1502,77 @@ def maybe_convert_dtype(data, copy):
     return data, copy
 
 
+def objects_to_datetime64ns(data, dayfirst, yearfirst,
+                            utc=False, errors="raise",
+                            require_iso8601=False, allow_object=False):
+    """
+    Convert data to array of timestamps.
+
+    Parameters
+    ----------
+    data : np.ndarray[object]
+    dayfirst : bool
+    yearfirst : bool
+    utc : bool, default False
+        Whether to convert timezone-aware timestamps to UTC
+    errors : {'raise', 'ignore', 'coerce'}
+    allow_object : bool
+        Whether to return an object-dtype ndarray instead of raising if the
+        data contains more than one timezone.
+
+    Returns
+    -------
+    result : ndarray
+        np.int64 dtype if returned values represent UTC timestamps
+        np.datetime64[ns] if returned values represent wall times
+        object if mixed timezones
+    inferred_tz : tzinfo or None
+
+    Raises
+    ------
+    ValueError : if data cannot be converted to datetimes
+    """
+    assert errors in ["raise", "ignore", "coerce"]
+
+    # if str-dtype, convert
+    data = np.array(data, copy=False, dtype=np.object_)
+
+    try:
+        result, tz_parsed = tslib.array_to_datetime(
+            data,
+            errors=errors,
+            utc=utc,
+            dayfirst=dayfirst,
+            yearfirst=yearfirst,
+            require_iso8601=require_iso8601
+        )
+    except ValueError as e:
+        try:
+            values, tz_parsed = conversion.datetime_to_datetime64(data)
+            # If tzaware, these values represent unix timestamps, so we
+            #  return them as i8 to distinguish from wall times
+            return values.view('i8'), tz_parsed
+        except (ValueError, TypeError):
```

Comment:
Presumably to re-raise the original exception if the fallback fails.  This is the existing behavior.

I think @mroeschke and I are vaguely planning to revisit this before long and combine datetime_to_datetime64 into array_to_datetime, fixing the many idiosyncracies of these calls.

Reply:
maybe i wasn't clear. I think you can simply remove the try/except and it will work the way it is now. (and same below).

Reply:
If the claim you're making is that it will raise under the same conditions, I agree.  If the claim is that it will raise _the same exception_, I don't.  i.e. if the `conversion.datetime_to_datetime64` were not inside its own try/except, then we could end up getting a TypeError in cases where we currently get a ValueError

Reply:
really? i don't think that is actually possible. the original exception is re-raised here.

Reply:
> the original exception is re-raised here

That's my point.  We are definitely not communicating well.  I'm going to take that as a queue to step away from the desk for a few minutes.

Reply:
OK, trying again.

My understanding of your suggestion is that you want to change:

```
try:
    result, inferred_tz = array_to_datetime(...)
except ValueError as e:
    try:
        result, inferred_tz = datetime_to_datetime64(...)
        return [...]
    except (ValueError, TypeError):
        raise e
```

to:

```
try:
    result, inferred_tz = array_to_datetime(...)
except ValueError as e:
    result, inferred_tz = datetime_to_datetime64(...)
    return [...]
```

My claim is that if we did that, then in cases where `datetime_to_datetime64` raises a `TypeError`, we would not re-raise `e` like we do in the status quo.

Reply:
@jreback did this resolve the miscommunication?

Reply:
again if datetime_to_datetime64 raises
won’t it just bubble up which is the same thing as a reraise here

Reply:
> That sounds like a "no" on the resolving miscommunication. Did I at least accurately summarize your suggested change?

not resolved, but I see from @TomAugspurger which what i was getting at. i guess ok for now.